## Contexto del juego

Los partidos de pádel registrados en este dataset corresponden a una variante recreativa
jugada por estudiantes universitarios. Las reglas de puntuación son las siguientes:

- El partido se juega en un formato de puntos directos (sin games ni sets).
- Gana el primer equipo en alcanzar **7 puntos** con una diferencia mínima de **2 puntos**.
- Si al llegar a 7 puntos no hay diferencia de 2, el partido continúa hasta
  que algún equipo la consiga.
- Cada punto es disputado por dos equipos de dos jugadores cada uno.
- Las posiciones de los jugadores en la cancha (LEFT/RIGHT, TOP/BOTTOM)
  pueden variar entre puntos dentro del mismo partido.

## EDA

**Base de Videos** (`Base_Videos_Final_202502.csv` / `Base_Videos_Final_202601.csv`)

| Variable | Tipo | Descripción |
|---|---|---|
| `frame_idx` | entero | Índice del frame dentro del video del punto |
| `duration` | decimal | Duración de cada frame en segundos (~0.0167 s = 60 fps) |
| `total_frames` | entero | Total de frames en el video del punto |
| `video_duration` | decimal | Duración total del video del punto en segundos |
| `player_id` | entero | Identificador numérico del jugador en el partido (1–4) |
| `player_name` | texto | Nombre completo del jugador |
| `team` | texto | Posición del equipo (top or bottom) |
| `zone` | texto | Zona de la cancha donde se ubica el jugador: *back_court*, *front_court*, etc. |
| `ball_position_x` | decimal | Posición horizontal de la pelota en la cancha (metros) |
| `ball_position_y` | decimal | Posición vertical de la pelota en la cancha (metros) |
| `player_position_x` | decimal | Posición horizontal del jugador en la cancha (metros) |
| `player_position_y` | decimal | Posición vertical del jugador en la cancha (metros) |
| `distance_ball_to_net` | decimal | Distancia de la pelota a la red (metros) |
| `distance_player_to_ball_m` | decimal | Distancia del jugador a la pelota (metros) |
| `distance_player_to_net_m` | decimal | Distancia del jugador a la red (metros) |
| `distance_player_to_teammate_m` | decimal | Distancia del jugador a su compañero de equipo (metros) |
| `player_hits_ball` | booleano | Si el jugador golpea la pelota en ese frame ('True'/'False' como texto) |
| `prev_x` | decimal | Posición X del jugador en el frame anterior (metros) |
| `prev_y` | decimal | Posición Y del jugador en el frame anterior (metros) |
| `player_displacement` | decimal | Desplazamiento del jugador respecto al frame anterior (metros) |
| `player_speed_mps` | decimal | Velocidad actual del jugador (metros/segundo) |
| `prev_speed` | decimal | Velocidad del jugador en el frame anterior (metros/segundo) |
| `player_acceleration_mps2` | decimal | Aceleración del jugador (metros/segundo²) |
| `ball_position_x_prev` | decimal | Posición X de la pelota en el frame anterior (metros) |
| `ball_position_y_prev` | decimal | Posición Y de la pelota en el frame anterior (metros) |
| `ball_displacement` | decimal | Desplazamiento de la pelota respecto al frame anterior (metros) |
| `ball_speed_mps` | decimal | Velocidad de la pelota (metros/segundo) |
| `time_since_last_hit` | decimal | Tiempo transcurrido desde el último golpe del jugador (segundos) |
| `partido` | entero | Identificador del partido al que pertenece el frame |
| `punto` | entero | Identificador del punto dentro del partido |

---

**Traza de Datos de Partidos** (`Traza_Datos_Partidos_202502.xlsx` / `_202601.xlsx` · 2 hojas: Cancha 1 y Cancha 2)

| Variable | Tipo | Descripción |
|---|---|---|
| `ID_PARTIDO` | entero | Identificador único del partido |
| `CANCHA` | entero | Número de la cancha donde se jugó (1 o 2) |
| `ID_PUNTO` | entero | Número de secuencia del punto dentro del partido |
| `ID_JUGADOR_1_EQUIPO_1` | texto | Nombre del jugador 1 del equipo 1 |
| `ID_JUGADOR_2_EQUIPO_1` | texto | Nombre del jugador 2 del equipo 1 |
| `ID_JUGADOR_1_EQUIPO_2` | texto | Nombre del jugador 1 del equipo 2 |
| `ID_JUGADOR_2_EQUIPO_2` | texto | Nombre del jugador 2 del equipo 2 |
| `EQUIPO_1_POSICION` | texto | Posición general del equipo 1 en la cancha: *BOTTOM* o *TOP* |
| `EQUIPO_2_POSICION` | texto | Posición general del equipo 2 en la cancha: *BOTTOM* o *TOP* |
| `POSICION_JUGADOR_1_EQUIPO_1` | texto | Lado de la cancha del jugador 1 equipo 1: *LEFT* o *RIGHT* |
| `POSICION_JUGADOR_2_EQUIPO_1` | texto | Lado de la cancha del jugador 2 equipo 1: *LEFT* o *RIGHT* |
| `POSICION_JUGADOR_1_EQUIPO_2` | texto | Lado de la cancha del jugador 1 equipo 2: *LEFT* o *RIGHT* |
| `POSICION_JUGADOR_2_EQUIPO_2` | texto | Lado de la cancha del jugador 2 equipo 2: *LEFT* o *RIGHT* |
| `PUNTO_EQUIPO_1` | entero | Si el equipo 1 ganó el punto (1 = ganó, 0 = perdió) |
| `PUNTO_EQUIPO_2` | entero | Si el equipo 2 ganó el punto (1 = ganó, 0 = perdió) |

---

**Datos de Jugadores** (`Datos_Jugadores_Padel_202502.xlsx` / `Data_Jugadores_202601.xlsx`)

| Variable | Tipo | Descripción |
|---|---|---|
| `Marca temporal` | fecha | Fecha y hora en que se diligenció el formulario |
| `NOMBRE` | texto | Nombre completo del jugador (clave de enlace con la traza de partidos) |
| `CODIGO_ESTUDIANTE` | texto | Código institucional del estudiante |
| `CIUDAD` | texto | Ciudad de residencia del jugador |
| `EDAD` | entero | Edad del jugador en años |
| `GENERO` | texto | Género del jugador |
| `PROGRAMA_ACADEMICO` | texto | Programa universitario que cursa |
| `SEMESTRE` | entero | Semestre académico actual |
| `EXPERIENCIA_PADEL` | texto | Si tiene experiencia previa en pádel (Sí/No) |
| `TIEMPO_JUGANDO_PADEL` | texto | Tiempo aproximado que lleva practicando pádel |
| `PRACTICA_OTRO_DEPORTE_RAQUETA` | texto | Si practica otro deporte de raqueta (tenis, squash, etc.) |
| `NIVEL_ACTUAL_PADEL` | texto | Nivel de juego autorreportado (principiante, intermedio, avanzado) |
| `ESTADO_FISICO` | texto | Percepción del jugador sobre su condición física actual |
| `FRECUENCIA_DEPORTE` | texto | Frecuencia semanal con la que practica deporte en general |
| `ESTATURA` | decimal | Estatura del jugador (verificar unidades: metros o centímetros) |
| `TALLA` | texto | Talla de ropa del jugador (S, M, L, XL, etc.) |
| `MANO_DOMINANTE` | texto | Mano dominante: derecha o izquierda |
| `RITMO_CARDIACO` | entero | Frecuencia cardíaca en reposo (ppm) |
| `LESIONES_LY` | texto | Si tuvo lesiones en el último año |
| `RAZON_PARTICIPACION` | texto | Motivo por el que participa en la actividad |
| `OBJETIVO_ACTIVIDAD` | texto | Objetivo personal con la actividad deportiva |
| `INTERES_PADEL` | texto | Nivel de interés en el pádel como deporte |
| `CLASES_PADEL` | texto | Si ha tomado clases formales de pádel |
| `DESAYUNO` | texto | Si desayunó el día de la actividad (Sí/No) |
| `¿Que desayunaste?` | texto | Descripción abierta de lo que desayunó |
| `GENERO_MUSICAL_FAVORITO` | texto | Género musical favorito (dato contextual) |
| `COMENTARIOS` | texto | Observaciones adicionales en texto libre |
| `Puntos_Ganados` | entero | Total de puntos ganados en todos sus partidos |
| `Puntos_Jugados` | entero | Total de puntos disputados |
| `Win_Rate` | decimal | Tasa de éxito: Puntos_Ganados / Puntos_Jugados (0 a 1) |

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import unicodedata
warnings.filterwarnings("ignore")

# Cargar el dataset
data_videos = pd.read_csv("./databases/2026-1/Base_Videos_Final_202601.csv")
data_partidos = pd.concat([
    pd.read_excel("./databases/2026-1/Traza_Datos_Partidos_202601.xlsx", sheet_name="CANCHA_1"),
    pd.read_excel("./databases/2026-1/Traza_Datos_Partidos_202601.xlsx", sheet_name="CANCHA_2")
], ignore_index=True)
data_jugadores = pd.read_excel("./databases/2026-1/Data_Jugadores_202601.xlsx")

In [2]:
def normalizar_nombre(nombre):
    if pd.isna(nombre):
        return nombre
    # Minúsculas, strip, title
    nombre = str(nombre).strip().lower().title()
    # Quitar tildes
    nombre = unicodedata.normalize('NFKD', nombre).encode('ascii', 'ignore').decode('utf-8')
    return nombre

# Normalizar en data_videos
data_videos['player_name'] = data_videos['player_name'].apply(normalizar_nombre)

# Normalizar en data_partidos
cols_jugadores = [
    'ID_JUGADOR_1_EQUIPO_1', 'ID_JUGADOR_2_EQUIPO_1',
    'ID_JUGADOR_1_EQUIPO_2', 'ID_JUGADOR_2_EQUIPO_2'
]
for col in cols_jugadores:
    data_partidos[col] = data_partidos[col].apply(normalizar_nombre)

# Normalizar en data_jugadores
data_jugadores['NOMBRE'] = data_jugadores['NOMBRE'].apply(normalizar_nombre)

In [3]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 1000)

### LIMPIEZA DE LOS DATOS

#### data_videos

##### Valores nulos

**Detección**

In [4]:
data_videos.info()

<class 'pandas.DataFrame'>
RangeIndex: 434198 entries, 0 to 434197
Data columns (total 30 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   frame_idx                      434198 non-null  int64  
 1   duration                       434198 non-null  float64
 2   total_frames                   434198 non-null  int64  
 3   video_duration                 434198 non-null  float64
 4   player_id                      434198 non-null  int64  
 5   player_name                    434198 non-null  str    
 6   team                           225038 non-null  str    
 7   zone                           434198 non-null  str    
 8   ball_position_x                434198 non-null  float64
 9   ball_position_y                434198 non-null  float64
 10  player_position_x              434198 non-null  float64
 11  player_position_y              434198 non-null  float64
 12  distance_ball_to_net           434198 non

**Hallazgos**

- Hay frames en los cuales el equipo no es detectado team (la posición del equipo top or bottom) -> Al no cambiar durante la duración del punto el partido, se reemplazara con la media de ese punto o partido
- Hay casos en los que la distancia al compañero del jugador esta calculada en solo uno de los dos miembros del equipo -> Revisar la existencia de esos casos y corregirlo; En caso de que no pero se tengan las posiciones de los dos jugadores calcularlo para amos
- Los valores nulos de las variables prev_x, prev_y, player_displacement, player_speed, prev_speed, player_acceleration, ball_position_x_prev, ball_position_y_prev, ball_displacement, ball_speed_mps, son nulos en los primeros frames de cada punto -> Se plantea eliminar las filas de estos primeros frames
- Los valores nulos de las variables ball_position_x_prev, ball_position_y_prev, ball_displacement, ball_speed_mps que no son el primer frame no son calculadas -> Revisar si se cuenta con las posiciones de la pelota y corregirlo
- time_since_last_hit es nulo si al principio de cada punto hasta que no se golpee por primera vez -> imputar con -1, indica que no ha sido golpeado todavia
- tras una primera limpieza de los primeros frames quedarón apenas 285 nulos en las variables de ball_position_x_prev, ball_position_y_prev, ball_displacement, ball_speed_mps -> al ser tan pocos se opto por su eliminación
- en la player_name aparecen nombres Player_X (con X siendo diferentes números), esta es la forma del programa de power vision de registrar los jugadores que no pudo reconocer -> imputar este nombre y el id del jugador con el id del punto, el partido, y por el frame id, ya que por cada punto y partido debería haber 4 frame con el mismo id, entonces por descarte se puede saber de que jugar es ese id, en caso de que no se pueda hacer por descarte (que en un frame no se detecten 4 jugadors) implica que en ese frame hay muy poca información y al no ser un número significativo de ellos (9716) se pueden eliminar
- para la variable zone, aparecen 3 valores left_box, right_box y back_court, pero los videos se tomaron siempre cuando los jugadores estaban dentro de los límites de la cancha por lo cual esta es la forma del programa de decir que no los esta detectando -> imputar por partido y id del jugador
- después de una primera iteración para imputar los datos de zone, hubieron 1275 registros nulos, en los cuales no hubo una moda con la cual imputar los datos -> al ser muy pocos frames del computo global, aunque se podrían haber eliminados, se decidio calcular la zona a la que suelen entrar, que aunque este errada no tendrá un gran impacto y se podrá conservar el resto de la información

**Tratamiento**

In [5]:
# ============================================================
# TRATAMIENTO DE VALORES NULOS - data_videos
# ============================================================

# ------------------------------------------------------------
# 1. Eliminar primeros frames de cada punto
#    (prev_x, prev_y, player_displacement, player_speed,
#     prev_speed, player_acceleration y variables de pelota
#     son nulos estructuralmente en el primer frame)
# ------------------------------------------------------------
data_videos = data_videos[data_videos['frame_idx'] != 0].reset_index(drop=True)

data_videos = data_videos.dropna(
    subset=['prev_speed', 'player_acceleration_mps2']
).reset_index(drop=True)

# ------------------------------------------------------------
# 2. Imputar time_since_last_hit con -1
#    (nulo hasta que ocurre el primer golpe del punto)
# ------------------------------------------------------------
data_videos['time_since_last_hit'] = data_videos['time_since_last_hit'].fillna(-1)

# ------------------------------------------------------------
# 3. Imputar team con moda por partido y jugador
#    (no cambia durante el punto/partido)
# ------------------------------------------------------------
def imputar_moda(series):
    moda = series.dropna().mode()
    return series.fillna(moda[0] if not moda.empty else np.nan)

# Nivel 1: moda por partido y jugador
data_videos['team'] = (
    data_videos
    .groupby(['partido', 'player_id'])['team']
    .transform(imputar_moda)
)

# Nivel 2: moda global por jugador
data_videos['team'] = (
    data_videos
    .groupby('player_name')['team']
    .transform(imputar_moda)
)

data_videos = data_videos.dropna(subset=['team']).reset_index(drop=True)

# ------------------------------------------------------------
# 4. Imputar player_name y player_id de Player_X por descarte
#    Lógica: en cada (partido, punto, frame_idx) deben aparecer
#    4 jugadores. Si hay exactamente 1 Player_X y 3 conocidos,
#    se puede identificar al desconocido por descarte.
# ------------------------------------------------------------
jugadores_por_partido = (
    data_partidos[[
        'ID_PARTIDO',
        'ID_JUGADOR_1_EQUIPO_1', 'ID_JUGADOR_2_EQUIPO_1',
        'ID_JUGADOR_1_EQUIPO_2', 'ID_JUGADOR_2_EQUIPO_2'
    ]]
    .drop_duplicates('ID_PARTIDO')
    .melt(id_vars='ID_PARTIDO', value_name='player_name')
    .drop(columns='variable')
    .rename(columns={'ID_PARTIDO': 'partido'})
)
jugadores_por_partido['player_name'] = jugadores_por_partido['player_name'].apply(normalizar_nombre)

corregidos_playerx = 0
excepciones_playerx = 0

mask_playerx = data_videos['player_name'].str.contains(r'^Player_\d+$', na=False)
frames_playerx = data_videos[mask_playerx][['partido', 'punto', 'frame_idx']].drop_duplicates()

for _, row in frames_playerx.iterrows():
    partido, punto, frame = row['partido'], row['punto'], row['frame_idx']

    mask_frame = (
        (data_videos['partido'] == partido) &
        (data_videos['punto'] == punto) &
        (data_videos['frame_idx'] == frame)
    )

    jugadores_frame = data_videos[mask_frame]
    playerx_en_frame = jugadores_frame[
        jugadores_frame['player_name'].str.contains(r'^Player_\d+$', na=False)
    ]

    # Solo corregir si hay exactamente 1 Player_X en el frame
    if len(playerx_en_frame) != 1:
        excepciones_playerx += len(playerx_en_frame)
        continue

    conocidos = jugadores_frame[
        ~jugadores_frame['player_name'].str.contains(r'^Player_\d+$', na=False)
    ]['player_name'].tolist()

    jugadores_partido = jugadores_por_partido[
        jugadores_por_partido['partido'] == partido
    ]['player_name'].tolist()

    faltante = [j for j in jugadores_partido if j not in conocidos]

    if len(faltante) == 1:
        idx = playerx_en_frame.index[0]
        data_videos.loc[idx, 'player_name'] = faltante[0]
        corregidos_playerx += 1
    else:
        excepciones_playerx += 1

print(f"Player_X corregidos por descarte: {corregidos_playerx}")
print(f"Player_X sin corrección (excepciones): {excepciones_playerx}")

# Diagnóstico de excepciones por partido
total_frames_playerx = mask_playerx.sum()
print(f"\nTotal frames con Player_X:     {total_frames_playerx:,}")
print(f"Porcentaje sobre total frames: {excepciones_playerx / len(data_videos) * 100:.2f}%")
print(f"Porcentaje sobre Player_X:     {excepciones_playerx / total_frames_playerx * 100:.1f}%")

excepciones_por_partido = (
    data_videos[
        mask_playerx &
        data_videos['player_name'].str.contains(r'^Player_\d+$', na=False)
    ]
    .groupby('partido')
    .size()
    .reset_index(name='frames_excepcion')
    .sort_values('frames_excepcion', ascending=False)
)
total_por_partido = (
    data_videos
    .groupby('partido')
    .size()
    .reset_index(name='total_frames')
)
excepciones_por_partido = excepciones_por_partido.merge(total_por_partido, on='partido')
excepciones_por_partido['pct'] = (
    excepciones_por_partido['frames_excepcion'] /
    excepciones_por_partido['total_frames'] * 100
).round(2)
print(f"\nDistribución de excepciones por partido:")
print(excepciones_por_partido.to_string(index=False))

# Eliminar frames de Player_X no corregibles
frames_antes = len(data_videos)
mask_playerx_residual = data_videos['player_name'].str.contains(r'^Player_\d+$', na=False)
data_videos = data_videos[~mask_playerx_residual].reset_index(drop=True)
frames_despues = len(data_videos)

print(f"\nFrames eliminados: {frames_antes - frames_despues:,}")
print(f"Shape final:       {data_videos.shape}")

# ------------------------------------------------------------
# 5. Imputar zone
#    - left_box / right_box → left / right
#    - back_court y nulos → moda de left/right por partido y jugador
#      Si no hay moda por partido, usar moda global por jugador
# ------------------------------------------------------------

# Normalizar left_box/right_box → left/right
data_videos['zone'] = data_videos['zone'].replace({
    'left_box':  'left',
    'right_box': 'right'
})

# Marcar back_court como NaN para imputar junto con los nulos reales
data_videos['zone'] = data_videos['zone'].replace({'back_court': np.nan})

def moda_zone(series):
    moda = series.dropna().mode()
    return moda[0] if not moda.empty else np.nan

# Nivel 1: moda por partido y jugador
moda_por_jugador = (
    data_videos
    .groupby(['partido', 'player_id'])['zone']
    .apply(moda_zone)
    .reset_index()
    .rename(columns={'zone': 'zone_moda'})
)

data_videos = data_videos.merge(moda_por_jugador, on=['partido', 'player_id'], how='left')

mask_nulo = data_videos['zone'].isna()
data_videos.loc[mask_nulo, 'zone'] = data_videos.loc[mask_nulo, 'zone_moda']
data_videos = data_videos.drop(columns='zone_moda')

# Nivel 2: moda global por jugador (para los que nunca tuvieron zona en ese partido)
moda_global = (
    data_videos
    .groupby('player_name')['zone']
    .apply(moda_zone)
    .reset_index()
    .rename(columns={'zone': 'zone_moda_global'})
)

data_videos = data_videos.merge(moda_global, on='player_name', how='left')

mask_nulo = data_videos['zone'].isna()
data_videos.loc[mask_nulo, 'zone'] = data_videos.loc[mask_nulo, 'zone_moda_global']
data_videos = data_videos.drop(columns='zone_moda_global')

# Nivel 3: eliminar si aún quedan nulos (jugador sin zona en ningún partido)
nulos_zone_residuales = data_videos['zone'].isna().sum()
if nulos_zone_residuales > 0:
    print(f"Eliminando {nulos_zone_residuales} filas con zone irrecuperable")
    data_videos = data_videos.dropna(subset=['zone']).reset_index(drop=True)

print(f"Zone — valores imputados nivel 1 (partido+jugador): {mask_nulo.sum()}")
print(f"Distribución zone:\n{data_videos['zone'].value_counts()}")
print(f"Nulos restantes en zone: {data_videos['zone'].isna().sum()}")

# ------------------------------------------------------------
# 6. Imputar distance_player_to_teammate_m
#    Si un miembro del equipo tiene la distancia y el otro no,
#    copiar el valor. Si ninguno la tiene pero se tienen las
#    posiciones, calcularla para ambos.
# ------------------------------------------------------------
data_videos = data_videos.sort_values(
    ['partido', 'punto', 'frame_idx', 'team', 'player_id']
).reset_index(drop=True)

grp_equipo = data_videos.groupby(['partido', 'punto', 'frame_idx', 'team'])

companero_x = grp_equipo['player_position_x'].transform(
    lambda g: g.iloc[::-1].values
)
companero_y = grp_equipo['player_position_y'].transform(
    lambda g: g.iloc[::-1].values
)
dist_companero_espejo = grp_equipo['distance_player_to_teammate_m'].transform(
    lambda g: g.iloc[::-1].values
)

distancia_calculada = np.sqrt(
    (data_videos['player_position_x'] - companero_x)**2 +
    (data_videos['player_position_y'] - companero_y)**2
)

# Primero: copiar del compañero si él la tiene
data_videos['distance_player_to_teammate_m'] = (
    data_videos['distance_player_to_teammate_m']
    .fillna(dist_companero_espejo)
)

# Segundo: calcular con posiciones si aún es nulo
data_videos['distance_player_to_teammate_m'] = (
    data_videos['distance_player_to_teammate_m']
    .fillna(distancia_calculada)
)

print(f"\nShape después del paso 6: {data_videos.shape}")

# ------------------------------------------------------------
# 7. Recalcular variables de pelota donde sean nulas
#    (frames que no son primeros pero tienen nulos)
# ------------------------------------------------------------
data_videos = data_videos.sort_values(
    ['partido', 'punto', 'player_id', 'frame_idx']
).reset_index(drop=True)

grp = data_videos.groupby(['partido', 'punto', 'player_id'])

data_videos['ball_position_x_prev'] = data_videos['ball_position_x_prev'].fillna(
    grp['ball_position_x'].shift(1)
)
data_videos['ball_position_y_prev'] = data_videos['ball_position_y_prev'].fillna(
    grp['ball_position_y'].shift(1)
)

ball_disp_calculado = np.sqrt(
    (data_videos['ball_position_x'] - data_videos['ball_position_x_prev'])**2 +
    (data_videos['ball_position_y'] - data_videos['ball_position_y_prev'])**2
)
data_videos['ball_displacement'] = data_videos['ball_displacement'].fillna(
    ball_disp_calculado
)
data_videos['ball_speed_mps'] = data_videos['ball_speed_mps'].fillna(
    data_videos['ball_displacement'] / data_videos['duration']
)

# Eliminar nulos residuales irrecuperables de pelota
nulos_pelota_antes = data_videos[['ball_position_x_prev', 'ball_position_y_prev',
                                   'ball_displacement', 'ball_speed_mps']].isnull().sum().sum()
data_videos = data_videos.dropna(subset=[
    'ball_position_x_prev', 'ball_position_y_prev',
    'ball_displacement', 'ball_speed_mps'
]).reset_index(drop=True)
print(f"\nFilas eliminadas por nulos residuales de pelota: {nulos_pelota_antes}")

# ------------------------------------------------------------
# Verificación final
# ------------------------------------------------------------
nulos = data_videos.isnull().sum()
nulos_restantes = nulos[nulos > 0]

if nulos_restantes.empty:
    print("\nSin nulos restantes")
else:
    print("\nNulos restantes por columna:")
    print(nulos_restantes)

print(f"\nShape final: {data_videos.shape}")

Player_X corregidos por descarte: 4499
Player_X sin corrección (excepciones): 9716

Total frames con Player_X:     14,215
Porcentaje sobre total frames: 2.25%
Porcentaje sobre Player_X:     68.4%

Distribución de excepciones por partido:
 partido  frames_excepcion  total_frames  pct
       4              1687         18164 9.29
       7              1356         32754 4.14
       9              1253         33690 3.72
      12              1247         13395 9.31
      32               637         13049 4.88
      24               491          8577 5.72
      15               432         16536 2.61
      19               393          9359 4.20
      17               354         17493 2.02
      21               325         17880 1.82
      28               320          6983 4.58
       3               298         20641 1.44
      30               259         10808 2.40
       5               219         37038 0.59
       1               177          6501 2.72
      27               133

##### Outliers

**Detección**

Para el detección de los outliers se realizó una comparación utilizando el rango inerquartilico y unos límitos físicos previamente definidos, esto debido a que varias variables como player_speed_mps, player_acceleration_mps2, ball_speed_mps, ball_displacement y player_displacement en su mayoría de valores son iguales a cero afectando su capacidad para definir unos límites inferiores y superiore fiables, como se puede ver a continuación:

In [6]:
# ============================================================
# DETECCIÓN DE OUTLIERS - Comparación IQR vs Límites Físicos
# ============================================================

vars_numericas = [
    'player_speed_mps', 'player_acceleration_mps2',
    'ball_speed_mps', 'ball_displacement',
    'distance_player_to_ball_m', 'distance_player_to_net_m',
    'distance_player_to_teammate_m', 'player_displacement'
]

limites_fisicos = {
    'player_speed_mps':              (0, 7),
    'player_acceleration_mps2':      (-10, 10),
    'ball_speed_mps':                (0, 50),
    'ball_displacement':             (0, 20),
    'distance_player_to_ball_m':     (0, 20),
    'distance_player_to_net_m':      (0, 10),
    'distance_player_to_teammate_m': (0, 10),
    'player_displacement':           (0, 2),
}

resultados = []

for col in vars_numericas:
    # --- IQR (3x) ---
    Q1 = data_videos[col].quantile(0.25)
    Q3 = data_videos[col].quantile(0.75)
    IQR = Q3 - Q1
    lim_iqr_inf = Q1 - 3 * IQR
    lim_iqr_sup = Q3 + 3 * IQR
    mask_iqr = (data_videos[col] < lim_iqr_inf) | (data_videos[col] > lim_iqr_sup)
    n_iqr = mask_iqr.sum()

    # --- Límites físicos ---
    min_fis, max_fis = limites_fisicos[col]
    mask_fis = (data_videos[col] < min_fis) | (data_videos[col] > max_fis)
    n_fis = mask_fis.sum()

    # --- Solo IQR / Solo físico / Ambos ---
    solo_iqr  = (mask_iqr & ~mask_fis).sum()
    solo_fis  = (mask_fis & ~mask_iqr).sum()
    ambos     = (mask_iqr & mask_fis).sum()

    resultados.append({
        'variable':          col,
        'min_real':          data_videos[col].min(),
        'max_real':          data_videos[col].max(),
        'lim_iqr_inf':       lim_iqr_inf,
        'lim_iqr_sup':       lim_iqr_sup,
        'lim_fis_inf':       min_fis,
        'lim_fis_sup':       max_fis,
        'outliers_iqr':      n_iqr,
        'outliers_iqr_%':    round(n_iqr / len(data_videos) * 100, 3),
        'outliers_fis':      n_fis,
        'outliers_fis_%':    round(n_fis / len(data_videos) * 100, 3),
        'solo_iqr':          solo_iqr,
        'solo_fisico':       solo_fis,
        'ambos':             ambos,
    })

df_comparacion = pd.DataFrame(resultados).set_index('variable')

pd.set_option('display.float_format', '{:.3f}'.format)
print("=== Comparación IQR vs Límites Físicos ===\n")
print(df_comparacion.to_string())

=== Comparación IQR vs Límites Físicos ===

                                min_real  max_real  lim_iqr_inf  lim_iqr_sup  lim_fis_inf  lim_fis_sup  outliers_iqr  outliers_iqr_%  outliers_fis  outliers_fis_%  solo_iqr  solo_fisico   ambos
variable                                                                                                                                                                                         
player_speed_mps                   0.000   562.860        0.000        0.000            0            7         73492          17.440          4928           1.169     68564            0    4928
player_acceleration_mps2      -33534.611 33534.611        0.000        0.000          -10           10        116966          27.756        113931          27.036      3035            0  113931
ball_speed_mps                     0.000   976.108        0.000        0.000            0           50         83472          19.808          2906           0.690     80566        

**Hallazgos**

Como se comentó, en varias de estar variables, los limites que definió el rango interquartilico fueron cero o incluso valores negativos en el caso de las distancias.

**Tratamiento**

In [7]:
# ============================================================
# TRATAMIENTO DE OUTLIERS - Clipeo por límites físicos
# ============================================================

for col, (min_val, max_val) in limites_fisicos.items():
    if col == 'ball_displacement':
        continue  # sin outliers reales
    data_videos[col] = data_videos[col].clip(lower=min_val, upper=max_val)

# Verificación
print("=== Verificación post-clipeo ===")
for col, (min_val, max_val) in limites_fisicos.items():
    real_min = data_videos[col].min()
    real_max = data_videos[col].max()
    ok = "OK" if real_min >= min_val and real_max <= max_val else "REVISAR"
    print(f"{ok} | {col}: [{real_min:.3f}, {real_max:.3f}]")

=== Verificación post-clipeo ===
OK | player_speed_mps: [0.000, 7.000]
OK | player_acceleration_mps2: [-10.000, 10.000]
OK | ball_speed_mps: [0.000, 50.000]
OK | ball_displacement: [0.000, 16.285]
OK | distance_player_to_ball_m: [0.040, 20.000]
OK | distance_player_to_net_m: [0.000, 10.000]
OK | distance_player_to_teammate_m: [0.000, 10.000]
OK | player_displacement: [0.000, 2.000]


### data_partidos

#### Valores nulos

**Detección**

In [8]:
print(data_partidos.shape)
data_partidos.info()

(406, 15)
<class 'pandas.DataFrame'>
RangeIndex: 406 entries, 0 to 405
Data columns (total 15 columns):
 #   Column                       Non-Null Count  Dtype
---  ------                       --------------  -----
 0   ID_PARTIDO                   406 non-null    int64
 1   CANCHA                       406 non-null    int64
 2   ID_PUNTO                     406 non-null    int64
 3   ID_JUGADOR_1_EQUIPO_1        406 non-null    str  
 4   ID_JUGADOR_2_EQUIPO_1        406 non-null    str  
 5   ID_JUGADOR_1_EQUIPO_2        406 non-null    str  
 6   ID_JUGADOR_2_EQUIPO_2        406 non-null    str  
 7   EQUIPO_1_POSICION            406 non-null    str  
 8   EQUIPO_2_POSICION            406 non-null    str  
 9   POSICION_JUGADOR_1_EQUIPO_1  406 non-null    str  
 10  POSICION_JUGADOR_2_EQUIPO_1  406 non-null    str  
 11  POSICION_JUGADOR_1_EQUIPO_2  406 non-null    str  
 12  POSICION_JUGADOR_2_EQUIPO_2  406 non-null    str  
 13  PUNTO_EQUIPO_1               406 non-null    int64


In [9]:
data_partidos.sample(20)

,ID_PARTIDO,CANCHA,ID_PUNTO,ID_JUGADOR_1_EQUIPO_1,ID_JUGADOR_2_EQUIPO_1,ID_JUGADOR_1_EQUIPO_2,ID_JUGADOR_2_EQUIPO_2,EQUIPO_1_POSICION,EQUIPO_2_POSICION,POSICION_JUGADOR_1_EQUIPO_1,POSICION_JUGADOR_2_EQUIPO_1,POSICION_JUGADOR_1_EQUIPO_2,POSICION_JUGADOR_2_EQUIPO_2,PUNTO_EQUIPO_1,PUNTO_EQUIPO_2
389,36,2,7,Laura Natalia Garcia,Cesar Camilo Diaz Cufino,Gabriel Alejandro Corzo Fernandez,Yeison David Nino Arias,TOP,BOTTOM,RIGHT,LEFT,LEFT,RIGHT,1,0
256,14,2,10,Juliana Gomez,Juan Sebastian Melo Gonzalez,Santiago Pena Beltran,Juan David Quitian,TOP,BOTTOM,RIGHT,LEFT,LEFT,RIGHT,1,0
324,26,2,6,Juan Esteban Bustos Marin,Yeison David Nino Arias,Jose Fabian Garcia,Luis Felipe Bonilla Patino,TOP,BOTTOM,LEFT,RIGHT,RIGHT,LEFT,1,0
131,25,1,8,Cesar Camilo Diaz Cufino,Laura Natalia Garcia,Sergio Gomez Osorio,Juan Sebastian Melo Gonzalez,BOTTOM,TOP,RIGHT,LEFT,RIGHT,LEFT,0,1
316,24,2,8,Nicolas Navarro D.,Juan Esteban Bustos Marin,Juan Felipe Mahecha,Carlos Felipe Andrade,TOP,BOTTOM,RIGHT,LEFT,LEFT,RIGHT,1,0
369,32,2,8,Jorge Garcia,Juliana Gomez,Sergio Gomez Osorio,Juan Guio,TOP,BOTTOM,RIGHT,LEFT,RIGHT,LEFT,1,0
46,9,1,5,Daniel Esteban Ortegon Jaramillo,Gabriel Alejandro Corzo Fernandez,Esteban Sequeda Henao,Juan Felipe Mahecha,BOTTOM,TOP,RIGHT,LEFT,RIGHT,LEFT,0,1
25,5,1,4,Isabella Buitrago,Samuel Francisco Rubio Casallas,Mateo Ramirez Cabrera,Maria Jose Palomino Carreno,TOP,BOTTOM,LEFT,RIGHT,LEFT,RIGHT,1,0
355,30,2,7,Juan Sebastian Melo Gonzalez,Juan David Quitian,Jorge Garcia,Juliana Gomez,TOP,BOTTOM,RIGHT,LEFT,RIGHT,LEFT,1,0
180,37,1,5,Juan Guio,Mateo Ramirez Cabrera,Juan David Quitian,Nicolas Navarro D.,BOTTOM,TOP,RIGHT,LEFT,RIGHT,LEFT,1,0


**Hallazgos**

No se encontró ningún dato nulo

#### Outliers

**Detección**

Todas las variables de esta base de datos son indices y variables categoricas. Teniendo esto en cuenta, lo que quedaba por revisar era el formato y las categorias de estas variables.

**Hallazgos**

- No se encontrarón valores que sobresalieran de las categorías que se esperaban. 
- Sim embago, los nombres de algunos participantes tenian un formato diferente al de los demás (todo en mayúsculas), por lo cual se procede a formatear estas celdas.

**Tratamiento**

In [10]:
# ============================================================
# FORMATEO DE NOMBRES DE JUGADORES - data_partidos
# ============================================================

cols_jugadores = [
    'ID_JUGADOR_1_EQUIPO_1',
    'ID_JUGADOR_2_EQUIPO_1',
    'ID_JUGADOR_1_EQUIPO_2',
    'ID_JUGADOR_2_EQUIPO_2'
]

for col in cols_jugadores:
    data_partidos[col] = data_partidos[col].str.strip().str.title()

# Verificación
print("=== Muestra post-formateo ===")
print(data_partidos[cols_jugadores].drop_duplicates().to_string())

=== Muestra post-formateo ===
                 ID_JUGADOR_1_EQUIPO_1              ID_JUGADOR_2_EQUIPO_1              ID_JUGADOR_1_EQUIPO_2              ID_JUGADOR_2_EQUIPO_2
0            Alejandra Salazar Galeano       Angel Santiago Toribio Godoy              Santiago Pena Beltran      Sahily Andrea Herrera Eljadue
11                    Juan David Rubio              Santiago Pena Beltran          Alejandra Salazar Galeano               Laura Natalia Garcia
22                   Isabella Buitrago    Samuel Francisco Rubio Casallas              Mateo Ramirez Cabrera        Maria Jose Palomino Carreno
33           Juan Esteban Bustos Marin                  Sofia Daza Garcia  Gabriel Alejandro Corzo Fernandez       Angel Santiago Toribio Godoy
42    Daniel Esteban Ortegon Jaramillo  Gabriel Alejandro Corzo Fernandez              Esteban Sequeda Henao                Juan Felipe Mahecha
51   Gabriel Alejandro Corzo Fernandez                          Pia Aroor           Cesar Camilo Diaz Cufi

### data_jugadores

In [11]:
data_jugadores.head()

,Marca temporal,NOMBRE,CODIGO_ESTUDIANTE,CIUDAD,EDAD,GENERO,PROGRAMA_ACADEMICO,SEMESTRE,EXPERIENCIA_PADEL,TIEMPO_JUGANDO_PADEL,PRACTICA_OTRO_DEPORTE_RAQUETA,NIVEL_ACTUAL_PADEL,ESTADO_FISICO,FRECUENCIA_DEPORTE,ESTATURA,TALLA,MANO_DOMINANTE,RITMO_CARDIACO,LESIONES_LY,RAZON_PARTICIPACION,OBJETIVO_ACTIVIDAD,INTERES_PADEL,CLASES_PADEL,DESAYUNO,¿Que desayunaste?,GENERO_MUSICAL_FAVORITO,COMENTARIOS,Puntos_Ganados,Puntos_Jugados,Win_Rate
0,2026-02-27 16:03:40.938,Juan Guio,353904,Cajica,18,Masculino,Ingeniería Informática,3,No,Menos de 6 meses,Tenis,Primera vez,Excelente,Más de 4 veces,170.000,60.000,Derecha,80,No,Me gusta el deporte,Jugar padel,Sí,Sí,Si,Huevos,Reggaeton,Felicidad,37.000,69.000,0.536
1,2026-02-27 16:04:55.951,Sofia Daza Garcia,332412,Bogota - Occidente,19,Femenino,Ingeniería Industrial,5,No,Menos de 6 meses,Otro,Primera vez,Regular,1-2 veces,170.000,72.000,Derecha,NaN,Sí,Clase,Aprender a jugar y relacionar los datos con un...,No,Sí,No,Nada,Reggaeton,Ninguno :),18.000,39.000,0.462
2,2026-02-27 16:05:14.289,Juan David Ortiz Rendon,350201,Chia,18,Masculino,Ingeniería Informática,3,No,Menos de 6 meses,No,Primera vez,Malo,0 veces,172.000,64.000,Derecha,No lo conozco,No,Proyecto probabilidad,Aprender de pádel y probabilidad,Sí,Tal vez,Si,Torta de banano,Rock,Ninguno,21.000,42.000,0.500
3,2026-02-27 16:05:39.778,Daniel Esteban Ortegon Jaramillo,273306,Bogota - Norte,23,Masculino,Administración de empresas,8,Sí,Menos de 6 meses,Tenis de mesa,Principiante,Bueno,Más de 4 veces,182.000,76.000,Derecha,NaN,Sí,Diversión y proyecto,Jugar y aprender,Sí,Sí,Si,Huevos,Reggaeton,NaN,33.000,71.000,0.465
4,2026-02-27 16:05:46.999,Juan David Rubio,352723,Chia,18,Masculino,Ingeniería Informática,3,No,Menos de 6 meses,Tenis de mesa,Primera vez,Excelente,Más de 4 veces,178.000,68.000,Derecha,80,No,Porque me pareció una actividad interesante,Aprender a jugar padel,No,No,Si,Huevos revueltos,Reggaeton,.,34.000,53.000,0.642


La base de datos de jugadores es la que se piensa tiene menos valor para el modelo que se pronone construir, de ella solo se van a utilizar las que pueden tener una relación más clara con el rendimiento de los jugadores como deportistas (TIEMPO_JUGANDO_PADEL, NIVEL_ACTUAL_ÑADEL, ESTADO_FÍSICO, etc), aplicandoles un ordinal encodigo. Adicionalmente, se van eliminar las columnas Puntos_Ganados, Puntos_Jugados y Win_Rate, ya que se pueden obtener de data_partidos de la cual se tiene mayor confianza. 


In [12]:
dj = data_jugadores.copy()

# ── 1. Eliminar columnas irrelevantes ────────────────────────
cols_drop = [
    'Marca temporal', 'CODIGO_ESTUDIANTE', 'CIUDAD', 'PROGRAMA_ACADEMICO',
    'SEMESTRE', 'RAZON_PARTICIPACION', 'OBJETIVO_ACTIVIDAD', 'INTERES_PADEL',
    'DESAYUNO', '¿Que desayunaste?', 'GENERO_MUSICAL_FAVORITO', 'COMENTARIOS',
    'MANO_DOMINANTE', 'RITMO_CARDIACO', 'Puntos_Ganados', 'Puntos_Jugados', 'Win_Rate'
]
dj = dj.drop(columns=cols_drop)

# ── 2. Encoding ordinal ──────────────────────────────────────
ordinal_maps = {
    'TIEMPO_JUGANDO_PADEL': {
        'Menos de 6 meses': 0,
        'Entre 6 meses y 1 año': 1,
        'Más de 1 año': 2
    },
    'NIVEL_ACTUAL_PADEL': {
        'Primera vez': 0,
        'Principiante': 1,
        'Intermedio': 2
    },
    'ESTADO_FISICO': {
        'Malo': 0,
        'Regular': 1,
        'Bueno': 2,
        'Excelente': 3
    },
    'FRECUENCIA_DEPORTE': {
        'No practico deportes': 0,
        '0 veces': 0,
        '1-2 veces': 1,
        '3-4 veces': 2,
        'Más de 4 veces': 3
    },
    'CLASES_PADEL': {
        'No': 0,
        'Tal vez': 1,
        'Sí': 2
    }
}

for col, mapping in ordinal_maps.items():
    dj[col] = dj[col].map(mapping)

# ── 3. Encoding binario ──────────────────────────────────────
dj['EXPERIENCIA_PADEL']            = (dj['EXPERIENCIA_PADEL'] == 'Sí').astype(int)
dj['LESIONES_LY']                  = (dj['LESIONES_LY'] == 'Sí').astype(int)
dj['PRACTICA_OTRO_DEPORTE_RAQUETA']= (dj['PRACTICA_OTRO_DEPORTE_RAQUETA'] != 'No').astype(int)

# ── 6. Verificación ──────────────────────────────────────────
print(f"\nShape final: {dj.shape}")
print(f"\nNulos:\n{dj.isnull().sum()}")
print(f"\nPrimeras filas:")
print(dj.head(10).to_string())

data_jugadores = dj.copy()


Shape final: (31, 13)

Nulos:
NOMBRE                           0
EDAD                             0
GENERO                           0
EXPERIENCIA_PADEL                0
TIEMPO_JUGANDO_PADEL             0
PRACTICA_OTRO_DEPORTE_RAQUETA    0
NIVEL_ACTUAL_PADEL               0
ESTADO_FISICO                    0
FRECUENCIA_DEPORTE               0
ESTATURA                         0
TALLA                            0
LESIONES_LY                      0
CLASES_PADEL                     0
dtype: int64

Primeras filas:
                             NOMBRE  EDAD     GENERO  EXPERIENCIA_PADEL  TIEMPO_JUGANDO_PADEL  PRACTICA_OTRO_DEPORTE_RAQUETA  NIVEL_ACTUAL_PADEL  ESTADO_FISICO  FRECUENCIA_DEPORTE  ESTATURA  TALLA  LESIONES_LY  CLASES_PADEL
0                         Juan Guio    18  Masculino                  0                     0                              1                   0              3                   3   170.000 60.000            0             2
1                 Sofia Daza Garcia 

#### Valores nulos

**Detección**

In [13]:
data_jugadores.info()

<class 'pandas.DataFrame'>
RangeIndex: 31 entries, 0 to 30
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   NOMBRE                         31 non-null     str    
 1   EDAD                           31 non-null     int64  
 2   GENERO                         31 non-null     str    
 3   EXPERIENCIA_PADEL              31 non-null     int64  
 4   TIEMPO_JUGANDO_PADEL           31 non-null     int64  
 5   PRACTICA_OTRO_DEPORTE_RAQUETA  31 non-null     int64  
 6   NIVEL_ACTUAL_PADEL             31 non-null     int64  
 7   ESTADO_FISICO                  31 non-null     int64  
 8   FRECUENCIA_DEPORTE             31 non-null     int64  
 9   ESTATURA                       31 non-null     float64
 10  TALLA                          31 non-null     float64
 11  LESIONES_LY                    31 non-null     int64  
 12  CLASES_PADEL                   31 non-null     int64  
dtypes: 

**Hallazgos**

- Todas las columnas de nuestro interes no tienen datos nulos

### Construcción de la variable objetivo y variables de contexto

> **Nota:** El partido con ID 23 finalizó con marcador 7-6 (diferencia de 1 punto),
> lo cual no cumple la regla de diferencia mínima. Se trata como un error de
> registro y no afecta el análisis.

In [14]:
# ============================================================
# CONSTRUCCIÓN DE LA VARIABLE OBJETIVO Y VARIABLES DE CONTEXTO
# ============================================================

data_partidos = data_partidos.sort_values(['ID_PARTIDO', 'ID_PUNTO']).reset_index(drop=True)

# ------------------------------------------------------------
# 1. Marcador acumulado (ya calculado, verificar que existe)
# ------------------------------------------------------------
data_partidos['MARCADOR_EQUIPO_1'] = (
    data_partidos.groupby('ID_PARTIDO')['PUNTO_EQUIPO_1'].cumsum()
)
data_partidos['MARCADOR_EQUIPO_2'] = (
    data_partidos.groupby('ID_PARTIDO')['PUNTO_EQUIPO_2'].cumsum()
)

# ------------------------------------------------------------
# 2. Ganador del partido (variable objetivo)
# ------------------------------------------------------------
ganador_partido = (
    data_partidos.groupby('ID_PARTIDO')
    .last()[['MARCADOR_EQUIPO_1', 'MARCADOR_EQUIPO_2']]
    .apply(lambda row: 1 if row['MARCADOR_EQUIPO_1'] > row['MARCADOR_EQUIPO_2'] else 2, axis=1)
    .rename('EQUIPO_GANADOR')
    .reset_index()
)

data_partidos = data_partidos.merge(ganador_partido, on='ID_PARTIDO')

# ------------------------------------------------------------
# 3. Variables de contexto del marcador
# ------------------------------------------------------------
data_partidos['DIFERENCIA_MARCADOR'] = (
    data_partidos['MARCADOR_EQUIPO_1'] - data_partidos['MARCADOR_EQUIPO_2']
)
data_partidos['PUNTOS_JUGADOS'] = (
    data_partidos['MARCADOR_EQUIPO_1'] + data_partidos['MARCADOR_EQUIPO_2']
)
data_partidos['WINRATE_ACUM_E1'] = (
    data_partidos['MARCADOR_EQUIPO_1'] / data_partidos['PUNTOS_JUGADOS']
)
data_partidos['WINRATE_ACUM_E2'] = (
    data_partidos['MARCADOR_EQUIPO_2'] / data_partidos['PUNTOS_JUGADOS']
)

# ------------------------------------------------------------
# 4. Variables de momentum
# ------------------------------------------------------------
# Racha de puntos consecutivos por equipo
def calcular_racha(serie):
    racha = []
    contador = 0
    for val in serie:
        if val == 1:
            contador += 1
        else:
            contador = 0
        racha.append(contador)
    return racha

data_partidos['RACHA_EQUIPO_1'] = (
    data_partidos.groupby('ID_PARTIDO')['PUNTO_EQUIPO_1']
    .transform(calcular_racha)
)
data_partidos['RACHA_EQUIPO_2'] = (
    data_partidos.groupby('ID_PARTIDO')['PUNTO_EQUIPO_2']
    .transform(calcular_racha)
)

# Puntos ganados en los últimos 5 puntos
data_partidos['PUNTOS_ULTIMOS_5_E1'] = (
    data_partidos.groupby('ID_PARTIDO')['PUNTO_EQUIPO_1']
    .transform(lambda x: x.rolling(5, min_periods=1).sum())
)
data_partidos['PUNTOS_ULTIMOS_5_E2'] = (
    data_partidos.groupby('ID_PARTIDO')['PUNTO_EQUIPO_2']
    .transform(lambda x: x.rolling(5, min_periods=1).sum())
)

# ------------------------------------------------------------
# Verificación
# ------------------------------------------------------------
print(data_partidos[[
    'ID_PARTIDO', 'ID_PUNTO',
    'MARCADOR_EQUIPO_1', 'MARCADOR_EQUIPO_2',
    'DIFERENCIA_MARCADOR', 'PUNTOS_JUGADOS',
    'RACHA_EQUIPO_1', 'RACHA_EQUIPO_2',
    'PUNTOS_ULTIMOS_5_E1', 'PUNTOS_ULTIMOS_5_E2',
    'WINRATE_ACUM_E1', 'WINRATE_ACUM_E2',
    'EQUIPO_GANADOR'
]].head(20).to_string())

    ID_PARTIDO  ID_PUNTO  MARCADOR_EQUIPO_1  MARCADOR_EQUIPO_2  DIFERENCIA_MARCADOR  PUNTOS_JUGADOS  RACHA_EQUIPO_1  RACHA_EQUIPO_2  PUNTOS_ULTIMOS_5_E1  PUNTOS_ULTIMOS_5_E2  WINRATE_ACUM_E1  WINRATE_ACUM_E2  EQUIPO_GANADOR
0            1         1                  1                  0                    1               1               1               0                1.000                0.000            1.000            0.000               1
1            1         2                  2                  0                    2               2               2               0                2.000                0.000            1.000            0.000               1
2            1         3                  2                  1                    1               3               0               1                2.000                1.000            0.667            0.333               1
3            1         4                  2                  2                    0               4     

### Validación - Posición del equipo

Ahora que se han agrupado, se quiere validar la posición (top/bottom; left/right) de los miembros del equipó contrastando ambas bases de datos, corrigiendolas dandole prioridad a data_partidos. 

In [15]:
    # ============================================================
    # LOOKUP CORREGIDO: incluye ID_PUNTO para partidos donde
    # la zona cambia entre puntos
    # ============================================================

    lookup = pd.melt(
        data_partidos[[
            'ID_PARTIDO', 'ID_PUNTO',
            'ID_JUGADOR_1_EQUIPO_1', 'ID_JUGADOR_2_EQUIPO_1',
            'ID_JUGADOR_1_EQUIPO_2', 'ID_JUGADOR_2_EQUIPO_2',
            'EQUIPO_1_POSICION', 'EQUIPO_2_POSICION',
            'POSICION_JUGADOR_1_EQUIPO_1', 'POSICION_JUGADOR_2_EQUIPO_1',
            'POSICION_JUGADOR_1_EQUIPO_2', 'POSICION_JUGADOR_2_EQUIPO_2'
        ]],
        id_vars=[
            'ID_PARTIDO', 'ID_PUNTO',
            'EQUIPO_1_POSICION', 'EQUIPO_2_POSICION',
            'POSICION_JUGADOR_1_EQUIPO_1', 'POSICION_JUGADOR_2_EQUIPO_1',
            'POSICION_JUGADOR_1_EQUIPO_2', 'POSICION_JUGADOR_2_EQUIPO_2'
        ],
        value_vars=[
            'ID_JUGADOR_1_EQUIPO_1', 'ID_JUGADOR_2_EQUIPO_1',
            'ID_JUGADOR_1_EQUIPO_2', 'ID_JUGADOR_2_EQUIPO_2'
        ],
        var_name='rol',
        value_name='player_name'
    )

    lookup['team_correcto'] = lookup.apply(
        lambda r: r['EQUIPO_1_POSICION'] if 'EQUIPO_1' in r['rol'] else r['EQUIPO_2_POSICION'],
        axis=1
    ).str.upper()

    lookup['zone_correcta'] = lookup.apply(
        lambda r: (
            r['POSICION_JUGADOR_1_EQUIPO_1'] if r['rol'] == 'ID_JUGADOR_1_EQUIPO_1' else
            r['POSICION_JUGADOR_2_EQUIPO_1'] if r['rol'] == 'ID_JUGADOR_2_EQUIPO_1' else
            r['POSICION_JUGADOR_1_EQUIPO_2'] if r['rol'] == 'ID_JUGADOR_1_EQUIPO_2' else
            r['POSICION_JUGADOR_2_EQUIPO_2']
        ),
        axis=1
    ).str.lower()

    lookup = lookup.rename(columns={
        'ID_PARTIDO': 'partido',
        'ID_PUNTO':   'punto'
    })[['partido', 'punto', 'player_name', 'team_correcto', 'zone_correcta']]

    # Merge por partido + punto + jugador
    data_videos = data_videos.merge(
        lookup, on=['partido', 'punto', 'player_name'], how='left'
    )

    mask_con_match = data_videos['team_correcto'].notna()

    # Diagnóstico de discrepancias ANTES de corregir
    discrepancias_team = (
        mask_con_match &
        (data_videos['team'] != data_videos['team_correcto'])
    ).sum()

    discrepancias_zone = (
        mask_con_match &
        data_videos['zone'].notna() &
        (data_videos['zone'] != data_videos['zone_correcta'])
    ).sum()

    total      = len(data_videos)
    corregidos = mask_con_match.sum()
    sin_match  = (~mask_con_match).sum()

    print(f"Total frames:                  {total:,}")
    print(f"Frames con match en lookup:    {corregidos:,} ({corregidos/total*100:.1f}%)")
    print(f"Sin match (Player_X u otros):  {sin_match:,} ({sin_match/total*100:.1f}%)")
    print(f"\nDiscrepancias en team:         {discrepancias_team:,} ({discrepancias_team/total*100:.2f}%)")
    print(f"Discrepancias en zone:         {discrepancias_zone:,} ({discrepancias_zone/total*100:.2f}%)")

    # Corregir team y zone
    data_videos.loc[mask_con_match, 'team'] = data_videos.loc[mask_con_match, 'team_correcto']
    data_videos.loc[mask_con_match, 'zone'] = data_videos.loc[mask_con_match, 'zone_correcta']

    data_videos = data_videos.drop(columns=['team_correcto', 'zone_correcta'])

    print(f"\nShape final: {data_videos.shape}")

Total frames:                  440,707
Frames con match en lookup:    421,578 (95.7%)
Sin match (Player_X u otros):  19,129 (4.3%)

Discrepancias en team:         421,578 (95.66%)
Discrepancias en zone:         284,819 (64.63%)

Shape final: (440707, 30)


**Hallazgos**

- El tratamiento realizado a team fue correcto, provocando que no hubiera ninguna discrepancia entre las dos bases de datos. Sin embargo, en cuanto zone hubo un total de discrepancias de 13,329 (3.20%) que tras analizar estos casos se identifico que se debian a los cambios de posición (left/right) que hicieron algunos equipos entre puntos y puntos.
- El shape aumentó de 421,406 a 440,707 — el merge por partido + punto + jugador está generando duplicados, probablemente porque en data_partidos hay múltiples filas por punto (una por cada punto jugado) y el melt las está expandiendo. -> Se plantea verificar si hay duplicados y eliminandolos que caso tal

In [16]:
# Verificar duplicados tras el merge
duplicados = data_videos.duplicated(
    subset=['partido', 'punto', 'frame_idx', 'player_id']
).sum()
print(f"Filas duplicadas: {duplicados:,}")

# Verificar si el lookup tiene duplicados por partido+punto+jugador
duplicados_lookup = lookup.duplicated(
    subset=['partido', 'punto', 'player_name']
).sum()
print(f"Duplicados en lookup: {duplicados_lookup:,}")

# Eliminar duplicados manteniendo la primera ocurrencia
data_videos = data_videos.drop_duplicates(
    subset=['partido', 'punto', 'frame_idx', 'player_id']
).reset_index(drop=True)

print(f"Shape tras eliminar duplicados: {data_videos.shape}")

Filas duplicadas: 19,301
Duplicados en lookup: 64
Shape tras eliminar duplicados: (421406, 30)


### Agrupación por punto

In [17]:
# ============================================================
# AGREGACIÓN DE data_videos A NIVEL DE PUNTO
# ============================================================

# Normalizar team a mayúsculas
data_videos['team'] = data_videos['team'].str.upper()

data_videos_agg = data_videos.groupby(['partido', 'punto']).agg(

    # --- Duración del punto ---
    duracion_punto                  = ('video_duration', 'first'),

    # --- Velocidad de jugadores por equipo ---
    velocidad_prom_e1               = ('player_speed_mps', lambda x: x[data_videos.loc[x.index, 'team'] == 'BOTTOM'].mean()),
    velocidad_prom_e2               = ('player_speed_mps', lambda x: x[data_videos.loc[x.index, 'team'] == 'TOP'].mean()),
    velocidad_max_e1                = ('player_speed_mps', lambda x: x[data_videos.loc[x.index, 'team'] == 'BOTTOM'].max()),
    velocidad_max_e2                = ('player_speed_mps', lambda x: x[data_videos.loc[x.index, 'team'] == 'TOP'].max()),

    # --- Desplazamiento total por equipo ---
    desplazamiento_total_e1         = ('player_displacement', lambda x: x[data_videos.loc[x.index, 'team'] == 'BOTTOM'].sum()),
    desplazamiento_total_e2         = ('player_displacement', lambda x: x[data_videos.loc[x.index, 'team'] == 'TOP'].sum()),

    # --- Distancia promedio a la red por equipo ---
    dist_red_prom_e1                = ('distance_player_to_net_m', lambda x: x[data_videos.loc[x.index, 'team'] == 'BOTTOM'].mean()),
    dist_red_prom_e2                = ('distance_player_to_net_m', lambda x: x[data_videos.loc[x.index, 'team'] == 'TOP'].mean()),

    # --- Distancia promedio al compañero por equipo ---
    dist_companero_prom_e1          = ('distance_player_to_teammate_m', lambda x: x[data_videos.loc[x.index, 'team'] == 'BOTTOM'].mean()),
    dist_companero_prom_e2          = ('distance_player_to_teammate_m', lambda x: x[data_videos.loc[x.index, 'team'] == 'TOP'].mean()),

    # --- Golpes por equipo ---
    hits_e1                         = ('player_hits_ball', lambda x: x[data_videos.loc[x.index, 'team'] == 'BOTTOM'].sum()),
    hits_e2                         = ('player_hits_ball', lambda x: x[data_videos.loc[x.index, 'team'] == 'TOP'].sum()),

    # --- Velocidad de la pelota ---
    velocidad_prom_pelota           = ('ball_speed_mps', 'mean'),
    velocidad_max_pelota            = ('ball_speed_mps', 'max'),

    # --- Aceleración promedio por equipo ---
    aceleracion_prom_e1             = ('player_acceleration_mps2', lambda x: x[data_videos.loc[x.index, 'team'] == 'BOTTOM'].mean()),
    aceleracion_prom_e2             = ('player_acceleration_mps2', lambda x: x[data_videos.loc[x.index, 'team'] == 'TOP'].mean()),

).reset_index()

print(data_videos_agg.shape)
print(data_videos_agg.head(10).to_string())

(287, 19)
   partido  punto  duracion_punto  velocidad_prom_e1  velocidad_prom_e2  velocidad_max_e1  velocidad_max_e2  desplazamiento_total_e1  desplazamiento_total_e2  dist_red_prom_e1  dist_red_prom_e2  dist_companero_prom_e1  dist_companero_prom_e2  hits_e1  hits_e2  velocidad_prom_pelota  velocidad_max_pelota  aceleracion_prom_e1  aceleracion_prom_e2
0        1      1           5.267              0.449              0.305             3.394             1.697                    2.319                    3.168             7.085             5.531                   3.880                   4.203        3        0                  1.313                50.000                0.000                0.032
1        1      2           9.500              0.444              0.846             7.000             7.000                    3.810                   68.175             6.840             5.310                   5.502                   3.938        3        7                  1.796              

### Join


#### data_videos_agg + data_partidos = data

In [19]:
    # ============================================================
    # JOIN: data_videos_agg + data_partidos
    # ============================================================

    data = data_videos_agg.merge(
        data_partidos,
        left_on=['partido', 'punto'],
        right_on=['ID_PARTIDO', 'ID_PUNTO'],
        how='inner'
    )

    # Eliminar columnas redundantes
    data = data.drop(columns=['ID_PARTIDO', 'ID_PUNTO'])

    cols_redundantes = [
        'EQUIPO_1_POSICION', 'EQUIPO_2_POSICION',
        'POSICION_JUGADOR_1_EQUIPO_1', 'POSICION_JUGADOR_2_EQUIPO_1',
        'POSICION_JUGADOR_1_EQUIPO_2', 'POSICION_JUGADOR_2_EQUIPO_2',
        'PUNTO_EQUIPO_1', 'PUNTO_EQUIPO_2'  # ya capturadas en MARCADOR y DIFERENCIA
    ]
    data = data.drop(columns=cols_redundantes)

    print(f"Shape data_videos_agg: {data_videos_agg.shape}")
    print(f"Shape data_partidos:   {data_partidos.shape}")
    print(f"Shape data (join):     {data.shape}")
    print(f"\nColumnas:\n{data.columns.tolist()}")
    print(f"\nPrimeras filas:")
    print(data.head())

Shape data_videos_agg: (287, 19)
Shape data_partidos:   (406, 26)
Shape data (join):     (283, 35)

Columnas:
['partido', 'punto', 'duracion_punto', 'velocidad_prom_e1', 'velocidad_prom_e2', 'velocidad_max_e1', 'velocidad_max_e2', 'desplazamiento_total_e1', 'desplazamiento_total_e2', 'dist_red_prom_e1', 'dist_red_prom_e2', 'dist_companero_prom_e1', 'dist_companero_prom_e2', 'hits_e1', 'hits_e2', 'velocidad_prom_pelota', 'velocidad_max_pelota', 'aceleracion_prom_e1', 'aceleracion_prom_e2', 'CANCHA', 'ID_JUGADOR_1_EQUIPO_1', 'ID_JUGADOR_2_EQUIPO_1', 'ID_JUGADOR_1_EQUIPO_2', 'ID_JUGADOR_2_EQUIPO_2', 'MARCADOR_EQUIPO_1', 'MARCADOR_EQUIPO_2', 'EQUIPO_GANADOR', 'DIFERENCIA_MARCADOR', 'PUNTOS_JUGADOS', 'WINRATE_ACUM_E1', 'WINRATE_ACUM_E2', 'RACHA_EQUIPO_1', 'RACHA_EQUIPO_2', 'PUNTOS_ULTIMOS_5_E1', 'PUNTOS_ULTIMOS_5_E2']

Primeras filas:
   partido  punto  duracion_punto  velocidad_prom_e1  velocidad_prom_e2  \
0        1      1           5.267              0.449              0.305   
1       

**Hallazgos**

- Se detecto que hay un gran problema con el punto 8, el id de todos sus puntos es igual a uno, se cree es un error del registro -> se propone reasignar los ids de ese partido

In [21]:
# ============================================================
# CORRECCIÓN PARTIDO 8: reasignar ID_PUNTO desde marcador
# ============================================================

# Solo el partido 8 tiene este bug
mask_p8 = data_partidos['ID_PARTIDO'] == 8

# Asignar punto_real = posición dentro del partido (1-based)
data_partidos.loc[mask_p8, 'ID_PUNTO'] = (
    data_partidos.loc[mask_p8]
    .groupby('ID_PARTIDO')
    .cumcount() + 1
)

# Verificar
print("Partido 8 tras corrección:")
print(
    data_partidos[mask_p8][
        ['ID_PARTIDO', 'ID_PUNTO', 'MARCADOR_EQUIPO_1', 'MARCADOR_EQUIPO_2']
    ].to_string(index=False)
)

# ============================================================
# REHACER EL JOIN con partido 8 corregido
# ============================================================

data = data_videos_agg.merge(
    data_partidos,
    left_on=['partido', 'punto'],
    right_on=['ID_PARTIDO', 'ID_PUNTO'],
    how='inner'
)

data = data.drop(columns=['ID_PARTIDO', 'ID_PUNTO'])

cols_redundantes = [
    'EQUIPO_1_POSICION', 'EQUIPO_2_POSICION',
    'POSICION_JUGADOR_1_EQUIPO_1', 'POSICION_JUGADOR_2_EQUIPO_1',
    'POSICION_JUGADOR_1_EQUIPO_2', 'POSICION_JUGADOR_2_EQUIPO_2',
    'PUNTO_EQUIPO_1', 'PUNTO_EQUIPO_2'
]
data = data.drop(columns=cols_redundantes)

# ============================================================
# VERIFICACIÓN FINAL
# ============================================================

print(f"\nShape final data: {data.shape}")

# Confirmar que partido 8 ahora tiene sus 10 puntos
p8_check = data[data['partido'] == 8][['partido', 'punto', 'MARCADOR_EQUIPO_1', 'MARCADOR_EQUIPO_2', 'EQUIPO_GANADOR']]
print("Partido 8 en data final:")
print(p8_check.to_string(index=False))

Partido 8 tras corrección:
 ID_PARTIDO  ID_PUNTO  MARCADOR_EQUIPO_1  MARCADOR_EQUIPO_2
          8         1                  1                  0
          8         2                  2                  0
          8         3                  3                  0
          8         4                  4                  0
          8         5                  5                  0
          8         6                  5                  1
          8         7                  5                  2
          8         8                  5                  3
          8         9                  6                  3
          8        10                  6                  4

Shape final data: (283, 35)
Partido 8 en data final:
 partido  punto  MARCADOR_EQUIPO_1  MARCADOR_EQUIPO_2  EQUIPO_GANADOR
       8      1                  1                  0               1
       8      2                  2                  0               1
       8      3                  3               

#### data += data_jugadores

Antes de hacer el join se esta realizando una verificación entre las bases de datos para saber si tenemos los mismos jugadores en ambas.

In [22]:
# Verificar pre-join
jugadores_data_v2 = set(
    data['ID_JUGADOR_1_EQUIPO_1'].tolist() +
    data['ID_JUGADOR_2_EQUIPO_1'].tolist() +
    data['ID_JUGADOR_1_EQUIPO_2'].tolist() +
    data['ID_JUGADOR_2_EQUIPO_2'].tolist()
)
jugadores_dj = set(data_jugadores['NOMBRE'].tolist())

sin_perfil = sorted(jugadores_data_v2 - jugadores_dj)
print(f"Jugadores sin perfil en data_jugadores ({len(sin_perfil)}):")
for j in sin_perfil:
    # Cuántos puntos tiene ese jugador en data
    n = sum(data[col].eq(j).sum() for col in [
        'ID_JUGADOR_1_EQUIPO_1','ID_JUGADOR_2_EQUIPO_1',
        'ID_JUGADOR_1_EQUIPO_2','ID_JUGADOR_2_EQUIPO_2'
    ])
    print(f"  {j}: aparece en {n} puntos")


Jugadores sin perfil en data_jugadores (9):
  Carlos Andres Diaz Mendez: aparece en 13 puntos
  Cesar Camilo Diaz Cufino: aparece en 39 puntos
  Jeronimo Sanabria Olivares: aparece en 13 puntos
  Juliana Gomez: aparece en 65 puntos
  Laura Natalia Garcia: aparece en 66 puntos
  Maria Jose Palomino Carreno: aparece en 9 puntos
  Natalia Perez: aparece en 10 puntos
  Nicolas Navarro D.: aparece en 17 puntos
  Sergio Gomez Osorio: aparece en 43 puntos


**Hallazgos**

- De los 36 jugadores que tiene data, data_jugadores solo tiene 9. -> Realizar imputación por mediana/moda o no utilizar la base de datos data_jugadores

#### Variables construidas

A continuación se describen las nuevas variables del dataframe data, las cuales se extrayeron de las bases de datos originales.

| Variable | Fuente | Tipo | Descripción |
|---|---|---|---|
| `duracion_punto` | video | decimal | Segundos desde el primer al último frame del punto |
| `velocidad_prom_e1` / `velocidad_prom_e2` | video | decimal | Velocidad media de los jugadores del equipo durante el punto (m/s) |
| `velocidad_max_e1` / `velocidad_max_e2` | video | decimal | Velocidad máxima registrada por cualquier jugador del equipo (m/s) |
| `desplazamiento_total_e1` / `desplazamiento_total_e2` | video | decimal | Suma de desplazamientos frame a frame de ambos jugadores del equipo (m) |
| `dist_red_prom_e1` / `dist_red_prom_e2` | video | decimal | Distancia promedio de los jugadores a la red durante el punto (m) |
| `dist_companero_prom_e1` / `dist_companero_prom_e2` | video | decimal | Distancia promedio entre compañeros de equipo durante el punto (m) |
| `hits_e1` / `hits_e2` | video | entero | Número total de golpes a la pelota realizados por el equipo en el punto |
| `velocidad_prom_pelota` | video | decimal | Velocidad media de la pelota durante el punto (m/s) |
| `velocidad_max_pelota` | video | decimal | Velocidad máxima registrada de la pelota durante el punto (m/s) |
| `aceleracion_prom_e1` / `aceleracion_prom_e2` | video | decimal | Aceleración promedio de los jugadores del equipo durante el punto (m/s²) |
| `MARCADOR_EQUIPO_1` / `MARCADOR_EQUIPO_2` | partidos | entero | Puntos acumulados por cada equipo hasta ese momento del partido |
| `EQUIPO_GANADOR` | partidos | entero | Equipo que ganó el punto (1 = equipo 1, 2 = equipo 2) |
| `DIFERENCIA_MARCADOR` | partidos | entero | Diferencia de puntos entre equipos al momento del punto (E1 − E2) |
| `PUNTOS_JUGADOS` | partidos | entero | Número de puntos disputados hasta ese momento en el partido |
| `WINRATE_ACUM_E1` / `WINRATE_ACUM_E2` | partidos | decimal | Proporción de puntos ganados por cada equipo hasta ese momento (0–1) |
| `RACHA_EQUIPO_1` / `RACHA_EQUIPO_2` | partidos | entero | Puntos consecutivos ganados por el equipo en ese momento |
| `PUNTOS_ULTIMOS_5_E1` / `PUNTOS_ULTIMOS_5_E2` | partidos | entero | Puntos ganados por cada equipo en los últimos 5 puntos disputados |